api/src/services/files.ts

```diff
@@ -65,39 +71,91 @@ export class FilesService extends ItemsService {
 			}
 		}
 
-		if (existingFile !== null && primaryKey !== undefined) {
-			await this.updateOne(primaryKey, payload, { emitEvents: false });
-
-			// If the file you're uploading already exists, we'll consider this upload a replace. In that case, we'll
-			// delete the previously saved file and thumbnails to ensure they're generated fresh
-			const disk = storage.location(payload.storage);
-			const filePrefix = path.parse(existingFile['filename_disk']).name;
+		// Is this file a replacement? if the file data already exists and we have a primary key
+		const isReplacement = existingFile !== null && primaryKey !== undefined;
 
-			for await (const filepath of disk.list(filePrefix)) {
-				await disk.delete(filepath);
-			}
-		} else {
+		// If this is a new file upload, we need to generate a new primary key and DB record
+		if (isReplacement === false || primaryKey === undefined) {
 			primaryKey = await this.createOne(payload, { emitEvents: false });
 		}
 
 		const fileExtension =
 			path.extname(payload.filename_download!) || (payload.type && '.' + extension(payload.type)) || '';
 
+		// The filename_disk is the FINAL filename on disk
 		payload.filename_disk = primaryKey + (fileExtension || '');
 
+		// Temp filename is used for replacements
+		const tempFilenameDisk = 'temp_' + payload.filename_disk;
+
 		if (!payload.type) {
 			payload.type = 'application/octet-stream';
 		}
 
+		// Used to clean up if something goes wrong
+		const cleanUp = async () => {
+			// If this is a new file upload that failed, we need to delete the DB record
+			if (isReplacement === false) {
+				await this.deleteOne(primaryKey!);
+			}
+
+			try {
+				if(isReplacement === true ){
+					// If this is a replacement that failed, we need to delete the temp file
+					await disk.delete(tempFilenameDisk);
+				} else {
+					// If this is a new file upload that failed, we need to delete the final file
+					await disk.delete(payload.filename_disk!);
+				}
+
+			} catch (err: any) {
+				logger.warn(`Couldn't delete temp file ${tempFilenameDisk}`);
+				logger.warn(err);
+			}
+		};
+
 		try {
-			await storage.location(data.storage).write(payload.filename_disk, stream, payload.type);
+			// If this is a replacement, we'll write the file to a temp location first to ensure we don't overwrite the existing file if something goes wrong
+			if (isReplacement === true) {
+				await disk.write(tempFilenameDisk, stream, payload.type);
+			} else {
+				// If this is a new file upload, we'll write the file to the final location
+				await disk.write(payload.filename_disk, stream, payload.type);
+			}
+
+			// Check if the file was truncated (if the stream ended early) and throw limit error if it was
+			// @ts-expect-error
```

Comment:
Lets type this in a way that doesn't disable TS

Reply:
I tried so many different ways but I don't know what was going on but I could not get it to type this right. Could use a look at this :)

Reply:
Some options I see that would resolve the error are:
1. Fix the function type (and any other dependant functions) to be FileStream only
2. Type assert the statement here (e.g. `(stream as FileStream).truncated`) as at this point we know it should have this property.
3. Add it as a property on all Readable Types
```ts
// e.g.
declare module 'stream' {
    interface Readable {
        truncated: boolean
    }
}
```

I would say 1 and 2 are our best options with 3 as a last resort as it will place that property on all Readable types.

Reply:
Some options I see that would resolve the error are:
1. Fix the function type (and any other dependant functions) to be FileStream only
2. Type assert the statement here (e.g. `(stream as FileStream).truncated`) as at this point we know it should have this property.
3. Add it as a property on all Readable Types
```ts
// e.g.
declare module 'stream' {
    interface Readable {
        truncated: boolean
    }
}
```

I would say 1 and 2 are our best options with 3 as a last resort as it will place that property on all Readable types.

Reply:
1. The type will not always be FileStream, not all file uploads go through BusBoy...importing via URL is just a normal Stream. Only streams through BusBoy have the truncated property added.
2. Stream will not always have that property, will this work?
3. BusBoy streams are not the only streams in the api, would this pollute other streams that don't have it?

Reply:
Seems type narrowing might be a better path. This works for me with no error
```ts
if ('truncated' in stream && stream.truncated === true) {
  // code
}
```